WIP: Add support for Wayland

jikortus opened the pull request:
**Part of the required functionality has been implemented so far:**
- Bundled `gnome-ponytail-daemon` binaries & friends.
- Bundled `ponytail` python module from `gnome-ponytail-daemon` project - manually (didn't make sense to me to bundle the whole git repo as a submodule just because of a single file).
- Added the wayland branch of dogtail as a git submodule.
- Added related services for `gnome-ponytail-daemon` and deployment of the binaries plus related bits in the environment.
- Added necessary environment variables (in systemd service files) + special handling in Anabot's launcher to avoid crashes or unnecessary warnings when running on X.
- The original hook used to wait for an X session has been replaced by a revamped one that also works with Wayland (also tightly related to the previous item).
- The last two commits are rather tentative and will likely be updated when RHEL-46789 (introspection support for `gnome-kiosk`) is implemented and I have a hands-on experience with how it works.

**Known issues:**
- We could potentially use `gnome-ponytail-daemon` binaries from Fedora, but I had no luck finding them for the secondary architectures.
- Errors related to missing `gsettings` tool in the log - I'll see how to handle this when there are the changes from RHEL-46789 implemented. If `gsettings` is not required, as the introspection should be IIUIC enabled by passing `--unsafe-mode` to `gnome-kiosk`, we could get rid of the errors by making dogtail happy with a dummy `gsettings` executable in `anabot-prepare.sh`. However, we may potentially need to file a RFE to stop removing `gsettings` in `lorax-templates-generic`.
- Duplicit messages for handler/check registration in the log. I haven't found the cause yet - it may likely not hurt anything, but is a bit annoying.
- We may want to rename a bit the dogtail/dogtail-wayland git submodule locations + modify their handling in the session hook when other changes are done and settled, but this is rather a cosmetic thing.

jikortus commented:
@jstodola As marked in the title, the whole thing is still WIP, but please feel free to share any related ideas or comments if you have any :slightly_smiling_face:

jikortus commented:
@jstodola I updated the summary to include all of the recent changes so that you can do a preliminary review.

jstodola commented on anabot/runtime/functions.py:
```diff
@@ -266,6 +266,8 @@ def getselected(parent, visible=True):
     return getnodes(parent, visible=visible, selected=True)
 
 def log_screenshot(wait=None, progress_only=False):
+# Temporarily disabled because of broken screenshot functionality in dogtail
```
Could this functionality be disabled just for wayland, so that we have screenshots on RHEL-9?

Also, could you add the RHEL issue number to the comments as well and fix the number it in the commit description? It should be RHEL-62420.

Also, s/RTT-5417/RTT-5714/ in the commit summary

jikortus replied:
That's a very good point, I was so submerged in the RHEL-10 stuff that I forgot to address this separately. I fixed that and added logging messages informing that the screenshot wasn't taken due to broken screenshot functionality on Wayland. OTOH it can get quite spammy, but I still think it's better than no information at all.

jstodola commented on etc/systemd/system/anabot.service.wants/xfr.service:
```diff
@@ -0,0 +1 @@
+../xfr.service
```
`xfr.service` doesn't exist

jikortus commented on anabot-prepare.sh:
```diff
@@ -3,3 +3,13 @@
+# (which means with '--unsafe-mode' parameter)
+echo "Creating wrapper script for /bin/gnome-kiosk to run in unsafe mode"
+mv /bin/gnome-kiosk /bin/gnome-kiosk_
+cat > /bin/gnome-kiosk << EOF
```
Thanks for reminding me of this possibility, fixed.

jikortus commented on etc/systemd/system/anabot-direct.service:
```diff
@@ -15,7 +15,7 @@ Environment="TERES_PATH=/opt/teres"
 Environment="XDG_RUNTIME_DIR=/run/user/0"
 # $WAYLAND_DISPLAY and $DBUS_SESSION_BUS_ADDRESS get undefined in anabot/launcher.py if running on X
-Environment="WAYLAND_DISPLAY=wayland-0"
+Environment="WAYLAND_DISPLAY=wl-sysinstall-0"
```
OK

jikortus commented on profiles/default/hooks/25-wait_for_session-preexec.hook:
```diff
@@ -12,20 +12,20 @@ waited=0
 while [ $waited -le $wait_timeout ]; do
     sleep 1
-    session_type=$(loginctl --no-legend)
+    session_type=$(loginctl show-session \
```
OK, done.

jikortus commented on profiles/default/hooks/25-wait_for_session-preexec.hook:
```diff
@@ -12,20 +12,20 @@ waited=0
     sleep 1
-    session_type=$(loginctl --no-legend)
+    session_type=$(loginctl show-session \
+        $(loginctl --no-legend | awk '{print $1}') -p Type | awk -F= '{print $2}' | \
```
I couldn't figure out how you meant it should work - I guess you thought the `-p Type` part belongs to `awk`, however it belongs to `loginctl`.

jikortus commented on debug_dogtail.py:
```diff
@@ -12,6 +12,18 @@
+with open("/run/anabot/session_type", "r") as st:
+    if st.read() == "wayland":
+        os.environ["XDG_RUNTIME_DIR"] = "/run/user/0"
+        os.environ["WAYLAND_DISPLAY"] = "wayland-0"
```
Sure, `wl-sysinstall-0` is correct, thank you!

jikortus commented:
@jstodola, thank you for the review. I've addressed all your findings (except the one with `awk`, although I think it might have confused you as the line in question is somewhat crammed).
Apart of that, there are a few other changes:
- Fix for the `gnome-kiosk` wrapper - originally the actual `gnome-kiosk` binary would get replaced by the wrapper upon restart, rendering it unusable. Even though I don't really thing restarts of anaconda/anabot would normally occur, it might be an unpleasant surprise when debugging some issues and actually forcing the restarts.
- I ensured proper dogtail commits are checked out in the respective (non)wayland dogtail directories (they weren't).
- The startup delay has been moved back from `python_launcher.sh` to the service file and will be used only with Xorg. The reason is that putting the delay in the launcher resulted in actually delaying every screenshot (as the screenshot functionality uses the launcher as well, which I didn't really realize initially, and had a fair bit of head-scratching later on). Also based on the observations I've made so far, it looks like removing the delay hasn't changed anything when Anabot is run on Wayland. I've also added a verbose comment to the service file and in the commit message, should we possibly need to implement the delay also in Wayland environments.
- The services dependencies have been fixed, so that the RDP client workaround service is started on s390x as well. However, I could see ~2 cases where this service and anabot-direct.service didn't get started, not really sure why. If you spot anything suspicious in the services, please let me know :-).

jikortus commented:
@jstodola, thank you for the review. I've addressed all your findings (except the one with `awk`, although I think it might have confused you as the line in question is somewhat crammed).
Apart of that, there are a few other changes:
- Fix for the `gnome-kiosk` wrapper - originally the actual `gnome-kiosk` binary would get replaced by the wrapper upon restart, rendering it unusable. Even though I don't really thing restarts of anaconda/anabot would normally occur, it might be an unpleasant surprise when debugging some issues and actually forcing the restarts.
- I ensured proper dogtail commits are checked out in the respective (non)wayland dogtail directories (they weren't).
- The startup delay has been moved back from `python_launcher.sh` to the service file and will be used only with Xorg. The reason is that putting the delay in the launcher resulted in actually delaying every screenshot (as the screenshot functionality uses the launcher as well, which I didn't really realize initially, and had a fair bit of head-scratching later on). Also based on the observations I've made so far, it looks like removing the delay hasn't changed anything when Anabot is run on Wayland. On
- The services dependencies have been fixed, so that the RDP client workaround service is started on s390x as well. However, I could see ~2 cases where this service and anabot-direct.service didn't get started, not really sure why. If you spot anything suspicious in the services, please let me know :-).

jstodola reviewed (approved):
It looks good to me, thank you and @You-never-know for all the effort to make anabot running with Wayland!
